### ISSUE TITLE
Feature request: Chained targets

### ISSUE THREAD

**Issue by kankaristo** [noseglid/atom-build#89]
Thanks for a great package, and thanks for putting in targets, they are extremely useful! :+1:

One more feature request (before I think of the next one :blush:): chained targets.

Essentially, this would be a target, which builds other targets. Let's say you have the targets:
- "build"
- "deploy"
- "run"

The "build" target just builds itself (nothing new here).

The "deploy" target builds the "build" target before building itself. If "build" fails, "deploy" won't run.

The "run" target builds the "deploy" target before building itself (and "deploy" builds the "build" target).

You could have "deploy A" and "deploy B", which would both run the same "build" target, etc.

You could also have some target always be run after "build". Say "run tests", without "run tests" having to depend on "build". So building "deploy" would run: "build", "run tests", "deploy".

You can already do all of this manually, but if you need to change the "build" target, you would also manually have to edit the "deploy" and "run" targets.

You could argue that this should be left to the actual build system. But chaining the targets with `atom-build` would allow you to see what failed. Instead of just saying "Build failed", it could also tell you if "Deployment failed" or "Run returned an error" (or more simply "Step failed: run").

This would make Atom a full-fledged IDE (apart from debugging, breakpoints, profiling, etc., but those are obviously out of scope for this package).

Do you think this is something that could be added to `atom-build`?

**Comment by kankaristo** [comment]
Another (perhaps simpler) way to look at this would be "sub-targets". In this, you would have a target which only contains other targets (doesn't have it's own command).

So you would have "build & deploy", which runs "build", then "deploy" (unless "build" failed).

This would probably be simpler to implement and use. Also, you could build "deploy", without having to build "build".

**Comment by noseglid** [comment]
I think this would make this package border to being a full-fledged build system in itself, which I do not intend it to be. I fell like build targets took the package along that path, but still it felt like a really useful feature of a package which is **executing** a build system.

If `atom-build` started supported chaining, then the next request would be dependency management and I feel like this is already accomplished by build systems such as `make`, `gulp` or `grunt`.

The best course of action I think would be to create a `Make` similar to this;

```Makefile
build:
    ...

deploy: build
    ...

run: deploy
    ...
```

Thanks for the suggestion though.

**Comment by kankaristo** [comment]
Alright, I fully understand. I believe in the "one tool, one task" approach.

How about being able to customize the the "Building..." and "Build failed" messages? If you're "building" a documentation or tests target, it would be nice to customize those on a per target basis.

And how about exporting commands for targets? So that you would have `build:build-targetname` for each target? This would allow a key binding for a specific target (like "run" or "deploy").

I could take a stab at these, if you think they're worth implementing. I'll send a pull request soon for better handling of environment variables, so that something like `PATH=$PATH:/added/path` works as expected.

**Comment by noseglid** [comment]
We have discussed the ability to register atom commands for the build targets in #37. I think it is a good idea... generally. However, I think it would require a neat way to bind key-commands to those targets as well, otherwise they would be quite useless. Since people have different targets in different projects I don't think the custom `keymaps.cson` would do. It's global.

Regarding the `PATH=$PATH:/added/path`. I think this should work since v0.35.0 (fixed in #69). Please let me know if this isn't the case!

**Comment by kankaristo** [comment]
I have v0.36.1 installed and `PATH=$PATH:/added/path` becomes `PATH=$PATH:/added/path:/added/path:/added/path`.

I'm not sure if this a regression or if it has always been like that.

I've fixed this, so that it becomes `PATH=<$PATH from process environment>:/added/path`, but I'd like to clean up the code a bit before I send in a PR.

----

About the per-target key bindings: you could add a class like `.build-targetname` to the `atom-workspace` element. Then, dynamically remove and add these classes for all available targets.

Maybe add another class for the project "name" (or something generated from the project path), so you could have a key binding for all targets named "target-a" **or** for just the target named "target-a" in the project named "project-a".

This *should* (I haven't tested it) allow for custom per-project and per-target key bindings in the user's `keymap.cson` with something like:
```
'atom-workspace.some-target':
  # Key binding for all targets named "some-target"
  'alt-q': 'build:build-some-target'
'atom-workspace.some-project.some-target':
  # Key binding for "some-target" in "some-project"
  'alt-w': 'build:build-some-target'
```

Personally, I would prefer setting the key bindings there, instead of in, say, the package's settings.
